internal/repository/azureblob.go

```diff
@@ -0,0 +1,246 @@
+		return
+	}
+
+	container = comps[1]
+	prefix = strings.Join(comps[2:], "/")
+
+	return
+}
+
+func makeAzureServiceAccountUrl(accountName string) string {
+	return fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)
+}
+
+func NewAzureBlobRepository(repo string) (*AzureBlobRepository, error) {
+	accountName, container, prefix, err := ParseAzureBlobUrl(repo)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceUrl := makeAzureServiceAccountUrl(accountName)
+	var serviceClient azblob.ServiceClient
+
+	var accountKey string
+	if value := os.Getenv("AZURE_STORAGE_ACCOUNT_KEY"); value != "" {
+		accountKey = value
+	}
+
+	if accountKey != "" {
+		credential, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceClient, err = azblob.NewServiceClientWithSharedKey(serviceUrl, credential, nil)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		credential, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceClient, err = azblob.NewServiceClient(serviceUrl, credential, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	containerClient := serviceClient.NewContainerClient(container)
+
+	r := &AzureBlobRepository{
+		Client:   &containerClient,
+		BasePath: repo,
+		Prefix:   prefix,
+	}
+
+	// check if the client has enough permission
+	dir, err := os.MkdirTemp("", "artivc-azblob-*")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	err = r.Download("refs/latest", filepath.Join(dir, "latest"), nil)
+	if err != nil {
+		var internalError *azblob.InternalError
+		if !errors.As(err, &internalError) {
+			return nil, err
+		}
+
+		var errStorage *azblob.StorageError
+		if !internalError.As(&errStorage) {
+			return nil, internalError
+		}
+
+		if errStorage.ErrorCode == azblob.StorageErrorCodeBlobNotFound {
+			// blob not found. but authentication/authorization check is ok. PASS
+		} else if errStorage.ErrorCode == azblob.StorageErrorCodeAuthorizationPermissionMismatch {
+			// authorization permission mismatch
+			log.Debugln(errStorage.Error())
+			fmt.Fprintf(os.Stderr, "Authorization permission mismatch. Please assign 'Storage Blob Data Contributor' role to the logged-in account in the storage account '%s'\n", accountName)
+			fmt.Fprintln(os.Stderr, "Please see https://docs.microsoft.com/azure/storage/blobs/assign-azure-role-data-access")
+			fmt.Fprintln(os.Stderr, "")
+			return nil, fmt.Errorf("authorization permission mismatch")
+		} else {
+			// other error
+			return nil, errStorage
+		}
+	}
+
+	return r, nil
+}
+
+func (repo *AzureBlobRepository) Upload(localPath, repoPath string, m *Meter) error {
+	ctx := context.Background()
+
+	// src
+	src, err := os.Open(localPath)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	// dest
+	blobPath := filepath.Join(repo.Prefix, repoPath)
+	blobClient := repo.Client.NewBlockBlobClient(blobPath)
+	blobClient.Upload(
```

Comment:
seems it could support parallelism?

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/storage/azblob@v0.3.0#BlockBlobClient.UploadFileToBlockBlob

```
blobClient.UploadFileToBlockBlob(
		ctx,
		src,
		azblob.HighLevelUploadToBlockBlobOption{Parallelism: 10})
```

Reply:
todo: meter support